systems/robotInterfaces/@Biped/getPelvisMotionForWalking.m

```diff
@@ -0,0 +1,218 @@
+
+pelvis_reference_height(1) = min(lsole_des(3),rsole_des(3));
+
+for i=1:length(support_times)-1
+  isDoubleSupport = any(supports(i).bodies==obj.foot_body_id.left) && any(supports(i).bodies==obj.foot_body_id.right);
+  isRightSupport = ~any(supports(i).bodies==obj.foot_body_id.left) && any(supports(i).bodies==obj.foot_body_id.right);
+  isLeftSupport = any(supports(i).bodies==obj.foot_body_id.left) && ~any(supports(i).bodies==obj.foot_body_id.right);
+  if options.debug
+    plot(support_times(i:i+1), 0.15 + 0.05*(isRightSupport|isDoubleSupport)*[1, 1], 'go:')
+    plot(support_times(i:i+1), 0.15 + 0.05*(isLeftSupport|isDoubleSupport)*[1,1], 'ro:')
+  end
+
+  nextIsDoubleSupport = any(supports(i+1).bodies==obj.foot_body_id.left) && any(supports(i+1).bodies==obj.foot_body_id.right);
+  nextIsRightSupport = ~any(supports(i+1).bodies==obj.foot_body_id.left) && any(supports(i+1).bodies==obj.foot_body_id.right);
+  nextIsLeftSupport = any(supports(i+1).bodies==obj.foot_body_id.left) && ~any(supports(i+1).bodies==obj.foot_body_id.right);
+
+  t = support_times(i);
+  t_next = support_times(i+1);
+
+  lfoot_frame = lfoot_body_motion.eval(t);
+  T_l_frame = poseQuat2tform([lfoot_frame(1:3); expmap2quat(lfoot_frame(4:6))]);
+  lsole_des = tform2poseQuat(T_l_frame * T_frame_to_sole.left);
+
+  rfoot_frame = rfoot_body_motion.eval(t);
+  T_r_frame = poseQuat2tform([rfoot_frame(1:3); expmap2quat(rfoot_frame(4:6))]);
+  rsole_des = tform2poseQuat(T_r_frame * T_frame_to_sole.right);
+
+  lfoot_frame_next = lfoot_body_motion.eval(t_next);
+  T_l_frame_next = poseQuat2tform([lfoot_frame_next(1:3); expmap2quat(lfoot_frame_next(4:6))]);
+  lsole_des_next = tform2poseQuat(T_l_frame_next * T_frame_to_sole.left);
+
+  rfoot_frame_next = rfoot_body_motion.eval(t_next);
+  T_r_frame_next = poseQuat2tform([rfoot_frame_next(1:3); expmap2quat(rfoot_frame_next(4:6))]);
+  rsole_des_next = tform2poseQuat(T_r_frame_next * T_frame_to_sole.right);
+
+  step_height_delta_threshold = 0.05; % cm, min change in height to classify step up/down
+  step_up_pelvis_shift = 0.07; % cm
+  if isDoubleSupport && nextIsDoubleSupport
+    pelvis_reference_height(i+1) = pelvis_reference_height(i);
+  elseif isDoubleSupport && nextIsLeftSupport
+    if lsole_des_next(3) > rsole_des(3) + step_height_delta_threshold
+      % stepping up with left foot
+      pelvis_reference_height(i+1) = lsole_des_next(3)-step_up_pelvis_shift;
+    else
+      pelvis_reference_height(i+1) = lsole_des_next(3);
+    end
+  elseif isDoubleSupport && nextIsRightSupport
+
+    if rsole_des_next(3) > lsole_des(3) + step_height_delta_threshold
+      % stepping up with right foot
+      pelvis_reference_height(i+1) = rsole_des_next(3)-step_up_pelvis_shift;
+    else
+      pelvis_reference_height(i+1) = rsole_des_next(3);
+    end
+  elseif isLeftSupport && nextIsDoubleSupport 
+    if rsole_des_next(3) < lsole_des(3) - step_height_delta_threshold
+      % stepping down with right foot
```

Comment:
this, on the other hand, looks correct to me (which makes me thing the logic above is in fact a mistake)?